pyispyb/app/extensions/database/definitions.py

```diff
@@ -70,3 +82,64 @@ def get_current_person(login: str) -> Optional[models.Person]:
     person._metadata["permissions"] = permissions
 
     return person
+
+
+def with_beamline_groups(
+    query: "sqlalchemy.orm.Query[Any]",
+    beamLineGroups: list[BeamLineGroup],
+    includeArchived: bool = False,
+    proposalColumn: "sqlalchemy.Column[Any]" = None,
+    joinBLSession: bool = True,
+    joinSessionHasPerson: bool = True,
+) -> "sqlalchemy.orm.Query[Any]":
+    """Apply beamline group based permissions
+
+    If the user is not a beamline group admin this will fallback to SessionHasPerson
+    """
+    # `all_proposals`` can access all sessions
+    if "all_proposals" in g.permissions:
+        logger.info("user has `all_proposals`")
+        return query
+
+    # Iterate through users permissions and match them to the relevant groups
+    beamLines = []
+    permissions_applied = []
+    for group in beamLineGroups:
+        if group.permission in g.permissions:
+            permissions_applied.append(group.permission)
+            for beamLine in group.beamLines:
+                if (beamLine.archived and includeArchived) or not includeArchived:
+                    beamLines.append(beamLine.beamLineName)
+
+    if proposalColumn:
+        query = query.join(
+            models.Proposal, models.Proposal.proposalId == proposalColumn
+        )
+
+    if joinBLSession:
+        query = query.join(
+            models.BLSession, models.BLSession.proposalId == models.Proposal.proposalId
+        )
+
+    if beamLines:
+        logger.info(
+            f"filtered to beamlines `{beamLines}` with permissions `{permissions_applied}`"
+        )
+
+        return query.filter(models.BLSession.beamLineName.in_(beamLines))
+    else:
```

Comment:
yes good point. i dont know how probable this is in reality (we never had a case like this in 10 years running synchweb), but i will have a think.

Reply:
Ok have attempted this one

Reply:
Looks great! Does it try to authorize through proposal_has_person too?

Also, maybe we should rename the method to `with_authorization` istead of `with_beamline_groups` since this is not only about beamline groups?

Some other thought:

- How about adding a `beamline_groups` field or method to the `g` object? This way we could simply use `g.beamline_groups` to know which groups the user can access just like we use `g.permissions`. That would avoid having to pass the list of groups as argument everywhere.


- Maybe we could simplify calling the authorization function by adding two others:
  - `with_proposal_authorization` for authorization on proposal-related entities:
    - joins with `proposal_has_person` and `Person`
    - **joins with `session`** and `session_has_person` and `Person` and checks beamline groups
  - `with_session_authorization` for authorization on session-related entities:
    - **joins with `Proposal`** and `proposal_has_person` and `Person`
    - joins with `session_has_person` and `Person` and checks beamline groups

Both check the same thing but with different join options so they would call the same function under the hood with different value presets for the boolean arguments `joinBLSession`, `joinSessionHasPerson`, (+ add `joinProposal`, `joinProposalHasPerson`?).
It would simplify the basic use case where we request a simple Proposal or BLSession entity and want to check authorization, while we could still use the main function with all the arguments for more complicated use cases.

What do you think?

Reply:
I'll look into renaming the function

> How about adding a beamline_groups field or method to the g object? This way we could simply use g.beamline_groups to know which groups the user can access just like we use g.permissions. That would avoid having to pass the list of groups as argument everywhere.

This is a nice idea, but unfortunately a bit complicated to get right based on the ordering of the request context and its associated middlewares. My initial try failed. I'd be happy to come back to this as a refactoring in a future mr.

> Maybe we could simplify calling the authorization function by adding two others:

`with_beamline_groups` should be pretty well configured for the default case i think. I needed the switches mostly for the awkward ones such as list of `sessions` and `datacollections`. I'd prefer to not add extra wrappers until they are really needed, theres only one function right now so no confusion about what to use. Heres the simple case:
https://github.com/ispyb/py-ispyb/pull/203/files#diff-b483ad95a687d98d32fffb1598af38c8a436baf24939144fb5665ab5ddc0edbcR40